stdlib/public/core/Array.swift

```diff
@@ -1865,3 +1865,168 @@ internal struct _ArrayAnyHashableBox<Element: Hashable>
+//   }
+//   return (result, pullback)
+// }
+// ```
+//
+extension Array : Differentiable where Element : Differentiable {
+  public struct TangentVector : AdditiveArithmetic & Differentiable {
+    public typealias TangentVector = Array<Element>.TangentVector
+    public typealias CotangentVector = Array<Element>.CotangentVector
+
+    public var elements: [Element.TangentVector]
+
+    public static var zero: TangentVector {
+      return TangentVector(elements: [])
+    }
+    
+    public init(elements: [Element.TangentVector]) {
+      self.elements = elements
+    }
+
+    @inlinable
+    public static func + (lhs: TangentVector, rhs: TangentVector) -> TangentVector {
+      return TangentVector(elements: zipLongest(lhs.elements, rhs.elements).map {
+        ($0 ?? .zero) + ($1 ?? .zero)
+      })
+    }
+
+    @inlinable
+    public static func - (lhs: TangentVector, rhs: TangentVector) -> TangentVector {
+      return TangentVector(elements: zipLongest(lhs.elements, rhs.elements).map {
+        ($0 ?? .zero) - ($1 ?? .zero)
+      })
+    }
+
+    @inlinable
+    public func moved(along direction: TangentVector) -> [Element.TangentVector] {
+      return zip(elements, direction.elements).map { $0.moved(along: $1) }
+    }
+
+    @inlinable
+    public func tangentVector(from cotangent: CotangentVector) -> TangentVector {
+      let elements = zip(self.elements, cotangent.elements).map {
+        $0.tangentVector(from: $1)
+      }
+      return TangentVector(elements: elements)
+    }
+  }
+
+  public struct CotangentVector : AdditiveArithmetic & Differentiable {
+    public typealias TangentVector = Array<Element>.CotangentVector
+    public typealias CotangentVector = Array<Element>.TangentVector
+
+    public var elements: [Element.CotangentVector]
+
+    public static var zero: TangentVector {
+      return TangentVector(elements: [])
+    }
+    
+    public init(elements: [Element.CotangentVector]) {
+      self.elements = elements
+    }
+
+    @inlinable
+    public static func + (lhs: TangentVector, rhs: TangentVector) -> TangentVector {
+      return TangentVector(elements: zipLongest(lhs.elements, rhs.elements).map {
+        ($0 ?? .zero) + ($1 ?? .zero)
+      })
+    }
+
+    @inlinable
+    public static func - (lhs: TangentVector, rhs: TangentVector) -> TangentVector {
+      return TangentVector(elements: zipLongest(lhs.elements, rhs.elements).map {
+        ($0 ?? .zero) - ($1 ?? .zero)
+      })
+    }
+
+    @inlinable
+    public func moved(along direction: TangentVector) -> [Element.CotangentVector] {
+      return zip(elements, direction.elements).map { $0.moved(along: $1) }
+    }
+
+    @inlinable
+    public func tangentVector(from cotangent: CotangentVector) -> TangentVector {
+      let elements = zip(self.elements, cotangent.elements).map {
+        $0.tangentVector(from: $1)
+      }
+      return TangentVector(elements: elements)
+    }
+  }
+
+  public typealias AllDifferentiableVariables = [Element]
```

Comment:
Put this on top along with `typealias TangentVector` and `typealias CotangentVector`?